Add `EMMREvaluator` and `MedianErrorEvaluator`

eukaryo opened the pull request:
<!-- Thank you for creating a pull request! In general, we merge your pull request after it gets two or more approvals. To proceed to the review process by the maintainers, please make sure that the PR meets the following conditions: (1) it passes all CI checks, and (2) it is neither in the draft nor WIP state. If you wish to discuss the PR in the draft state or need any other help, please mention the Optuna development team in the PR. -->

## Motivation
<!-- Describe your motivation why you will submit this PR. This is useful for reviewers to understand the context of PR. -->

I want to add new terminator based on "Expected Minimum Model Regret(EMMR)", which is an upper bound of "Expected Minimum Simple Regret" introduced in [the paper](https://proceedings.mlr.press/v206/ishibashi23a/ishibashi23a.pdf).

## Description of the changes
<!-- Describe the changes in this PR. -->

- Added two classes, `EMMREvaluator` and `MedianErrorEvaluator` in `optuna/terminator` module.
- Added unittests.
- Done minimal refactoring to `RegretBoundEvaluator` to call its methods from `EMMREvaluator`. Such calling would be justified from a theoretical perspective.

HideakiImamura commented on optuna/terminator/median_erroreval.py:
```diff
@@ -0,0 +1,78 @@
+            return -sys.float_info.max  # Do not terminate.
+        trials.sort(key=lambda trial: trial.number)
+        criteria = []
+        for i in range(self._warm_up_trials, self._warm_up_trials + self._n_initial_trials):
+            criteria.append(
+                self._paired_improvement_evaluator.evaluate(
+                    trials[self._warm_up_trials : self._warm_up_trials + i + 1], study_direction
+                )
+            )
```
```suggestion
        for i in range(1, self._n_initial_trials+1):
            criteria.append(
                self._paired_improvement_evaluator.evaluate(
                    trials[self._warm_up_trials : self._warm_up_trials + i], study_direction
                )
            )
```

HideakiImamura commented on tests/terminator_tests/test_median_erroreval.py:
```diff
@@ -0,0 +1,74 @@
+def test_ratio_to_initial_median_evaluator(
+    direction: str,
+) -> None:
+    study = create_study(direction=direction)
```
How about just preserving the list of trials?

HideakiImamura reviewed (commented):
Thanks for the PR. I have several comments. PTAL.

eukaryo commented on optuna/terminator/improvement/emmr.py:
```diff
@@ -0,0 +1,399 @@
+            normalized_params[:-1, :],
+            standarized_score_vals[:-1],
+            theta_t_star,
+            kernel_params_t1,
```
I reconsidered and realized that `mu_t2` is not needed. So I simply removed it and revised the affected code.

eukaryo commented on optuna/terminator/improvement/evaluator.py:
```diff
@@ -105,12 +93,28 @@ def _get_top_n(
+        return beta
+
+    @staticmethod
     def _compute_standardized_regret_bound(
```
I removed initial `_`, but remain it as a staticmethod. Since this method is tied to the paper on which `RegretBoundEvaluator` is based, it is natural to leave it as a staticmethod.

HideakiImamura reviewed (approved):
Thanks for the update. LGTM.

toshihikoyanase commented on optuna/terminator/improvement/emmr.py:
```diff
@@ -0,0 +1,390 @@
+
+        .. testcode::
+
+            import logging
```
Not a strong opinion, but we can use `print` instead of `logging.info` to simplify the example.

toshihikoyanase replied:
Please refer to the [code convention](https://github.com/optuna/optuna/wiki/Coding-Style-Conventions#block-and-inline-comments-are-supposed-to-be-sentences) of this repository.

eukaryo replied:
Agreed, but using `print` causes doctest failure. I have changed it to comment.

toshihikoyanase commented on optuna/terminator/improvement/emmr.py:
```diff
@@ -0,0 +1,390 @@
+        len_params = len(search_space.scale_types)
+        assert normalized_params.shape == (len_trials, len_params)
+
+        # _gp module assumes that optimization direction is maximization
```
A nit.
```suggestion
        # _gp module assumes that optimization direction is maximization.
```

toshihikoyanase replied:
Please refer to the [code convention](https://github.com/optuna/optuna/wiki/Coding-Style-Conventions#block-and-inline-comments-are-supposed-to-be-sentences) of this repository.

HideakiImamura commented:
I checked that the following script but encountered an overflow and got an empty figure. Could you take a look?

```python
import optuna
from optuna.terminator.erroreval import report_cross_validation_scores
from optuna.terminator import EMMREvaluator
from optuna.terminator import MedianErrorEvaluator
from optuna.visualization import plot_terminator_improvement

from sklearn.datasets import load_wine
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold


def objective(trial):
    X, y = load_wine(return_X_y=True)

    clf = RandomForestClassifier(
        max_depth=trial.suggest_int("max_depth", 2, 32),
        min_samples_split=trial.suggest_float("min_samples_split", 0, 1),
        criterion=trial.suggest_categorical("criterion", ("gini", "entropy")),
    )

    scores = cross_val_score(clf, X, y, cv=KFold(n_splits=5, shuffle=True))
    report_cross_validation_scores(trial, scores)
    return scores.mean()


if __name__ == "__main__":
    study = optuna.create_study()
    study.optimize(objective, n_trials=50)

    improvement_evaluator = EMMREvaluator()
    error_evaluator = MedianErrorEvaluator(improvement_evaluator)

    fig = plot_terminator_improvement(
        study, 
        plot_error=True, 
        improvement_evaluator=improvement_evaluator, 
        error_evaluator=error_evaluator,
    )
    fig.write_image("test.png")

```

eukaryo commented on optuna/terminator/median_erroreval.py:
```diff
@@ -0,0 +1,82 @@
+            The ``improvement_evaluator`` instance which is set with this ``error_evaluator``.
+        warm_up_trials:
+            A parameter specifies the number of initial trials to be discarded before
+            the calculation of median. Default to 10.
```
TODO: add comment

not522 commented on optuna/terminator/improvement/emmr.py:
```diff
@@ -0,0 +1,387 @@
+        theorem1_v = math.sqrt(
+            max(
+                1e-100,  # Note: sys.float_info.min can cause overflow in scipy_stats.
+                sigma_t_theta_t_star**2
```
I think this sigma is the variance, so isn't `**2` wrong?

not522 replied:
I think that the following diff resolves this problem.

```bash
~/optuna$ git diff
diff --git a/optuna/terminator/improvement/emmr.py b/optuna/terminator/improvement/emmr.py
index 0361ed8bf..6ac3c572e 100644
--- a/optuna/terminator/improvement/emmr.py
+++ b/optuna/terminator/improvement/emmr.py
@@ -241,9 +241,9 @@ class EMMREvaluator(BaseImprovementEvaluator):
         theorem1_v = math.sqrt(
             max(
                 1e-100,  # Note: sys.float_info.min can cause overflow in scipy_stats.
-                sigma_t_theta_t_star**2
+                sigma_t_theta_t_star
                 - 2.0 * cov_t_between_theta_t_star_and_theta_t1_star
-                + sigma_t_theta_t1_star**2,
+                + sigma_t_theta_t1_star,
             )
         )
         theorem1_g = (mu_t_theta_t_star - mu_t1_theta_t1_star) / theorem1_v
@@ -252,15 +252,15 @@ class EMMREvaluator(BaseImprovementEvaluator):
         alg1_delta_r_tilde_t_term3 = theorem1_v * theorem1_g * scipy_stats.norm.cdf(theorem1_g)
 
         _lambda = prior.DEFAULT_MINIMUM_NOISE_VAR**-1
-        eq4_rhs_term1 = 0.5 * math.log(1.0 + _lambda * sigma_t1_theta_t_with_nu_t**2)
+        eq4_rhs_term1 = 0.5 * math.log(1.0 + _lambda * sigma_t1_theta_t_with_nu_t)
         eq4_rhs_term2 = (
-            -0.5 * sigma_t1_theta_t_with_nu_t**2 / (sigma_t1_theta_t_with_nu_t**2 + _lambda**-1)
+            -0.5 * sigma_t1_theta_t_with_nu_t / (sigma_t1_theta_t_with_nu_t + _lambda**-1)
         )
         eq4_rhs_term3 = (
             0.5
-            * sigma_t1_theta_t_with_nu_t**2
+            * sigma_t1_theta_t_with_nu_t
             * (y_t - mu_t1_theta_t_with_nu_t) ** 2
-            / (sigma_t1_theta_t_with_nu_t**2 + _lambda**-1) ** 2
+            / (sigma_t1_theta_t_with_nu_t + _lambda**-1) ** 2
         )
 
         alg1_delta_r_tilde_t_term4 = kappa_t1 * math.sqrt(
```

not522 replied:
I tried to reproduce the figure 4 in the paper.
With the above fix, the results are closer to those in the paper.

```python
import math
import matplotlib.pyplot as plt
import optuna
from optuna.terminator import EMMREvaluator


def holder_table(trial):
    x = trial.suggest_float("x", -10, 10)
    y = trial.suggest_float("y", -10, 10)
    return -abs(math.sin(x) * math.cos(y) * math.exp(abs(1 - math.sqrt(x**2 + y**2) / math.pi)))

def cross_in_tray(trial):
    x = trial.suggest_float("x", -10, 10)
    y = trial.suggest_float("y", -10, 10)
    return -0.0001 * (abs(math.sin(x) * math.sin(y) * math.exp(abs(100 - math.sqrt(x**2 + y**2) / math.pi))) + 1) ** 0.1

def easom(trial):
    x = trial.suggest_float("x", -10, 10)
    y = trial.suggest_float("y", -10, 10)
    return -math.cos(x) * math.cos(y) * math.exp(-((x - math.pi)**2 + (y - math.pi)**2))

def six_hump_camel(trial):
    x = trial.suggest_float("x", -10, 10)
    y = trial.suggest_float("y", -10, 10)
    return (4 - 2.1 * x**2 + x**4 / 3) * x ** 2 + x * y + (-4 + 4 * y**2) * y**2

def rosenbrock(trial):
    a = 1
    b = 100
    x = trial.suggest_float("x", -10, 10)
    y = trial.suggest_float("y", -10, 10)
    return (a - x) ** 2 + b * (y - x ** 2) ** 2

def booth(trial):
    x = trial.suggest_float("x", -10, 10)
    y = trial.suggest_float("y", -10, 10)
    return (x + 2 * y - 7) ** 2 + (2 * x + y - 5) ** 2


for objective, index, name in (
        (holder_table, "a", "holder table"),
        (cross_in_tray, "b", "cross in tray"),
        (easom, "c", "easom"),
        (six_hump_camel, "d", "six hump camel"),
        (rosenbrock, "e", "rosenbrock"),
        (booth, "f", "booth")):
    x = []
    y1 = []
    y2 = []
    y3 = []
    y4 = []
    y5 = []

    sampler = optuna.samplers.TPESampler(seed=0)
    study = optuna.create_study(sampler=sampler, study_name=name)
    emmr_improvement_evaluator = EMMREvaluator()

    for i in range(500):
        trial = study.ask()
        value = objective(trial)
        study.tell(trial, value)

        print(i, study.best_value, value, trial.params)

        ret = emmr_improvement_evaluator.evaluate(
            trials=study.trials,
            study_direction=study.direction,
        )

        if isinstance(ret, tuple):
            t1, t2, t3, t4, t5 = ret
            x.append(i + 1)
            y1.append(t1)
            y2.append(t2)
            y3.append(t3)
            y4.append(t4)
            y5.append(t5)

    plt.plot(x, y1, ".", label="abs(term_1)")
    plt.plot(x, y2, ".", label="term_2+term_3")
    plt.plot(x, y3, ".", label="term_4")
    plt.plot(x, y4, ".", label="kappa")
    plt.plot(x, y5, ".", label="term_4/kappa")
    plt.yscale("log")
    plt.ylim(1e-12, 1000)
    plt.legend()
    plt.title(f"({index}) {name}")
    plt.savefig(f"emmr_{index}_{name.replace(' ', '_')}.png")
    plt.clf()
```

I modified `emmr_improvement_evaluator.evaluate` to return the terms of the evaluation.

```python
        return (
            abs(alg1_delta_r_tilde_t_term1),
            alg1_delta_r_tilde_t_term2+alg1_delta_r_tilde_t_term3,
            alg1_delta_r_tilde_t_term4,
            kappa_t1,
            alg1_delta_r_tilde_t_term4/kappa_t1,
        )
```

* current PR
![emmr_e_rosenbrock](https://github.com/user-attachments/assets/8a2a96ee-c103-4ca0-9020-fe07cb5bdbe0)

* fixed
![emmr_e_rosenbrock](https://github.com/user-attachments/assets/393d43cd-9626-4546-885d-e2a1fa52195d)

eukaryo commented:
- I have modified outputs and stopped using `sys.float_info.max`. I have confirmed that visualization works in current PR code.
- I have renamed the prefix `sigma_` to `variance_`, and removed `**2`.

not522 reviewed (approved):
Thanks for the update. I think merging this PR is fine.
@HideakiImamura Could you take another look? If it is OK, please merge this PR.

HideakiImamura reviewed (approved):
LGTM. Thanks for the long running work!
